### PR TITLE
Support both use_calc_stream and sync_op in all_reduce

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR types

New features

### PR changes

APIs

### Describe

In the new communication library, we designed ProcessGroup to manage different communication group. Inside each process_group has its own stream which all communications in this group will be done on this stream. For high level API, like distributed.all_reduce, we use `use_calc_stream` to indicate whether this operation is sync or not. Notice that frequently add unnecessary cuda events may lead to low performance on some model. In order to achieve high performance, this pr add a new API name distributed.stream.all_reduce. This new API provided `use_calc_stream` and `sync_op` both.
- `sync_op`, indicate whether communication is sync or not.
- `use_calc_stream`, do communicate on calc stream, save the time of switching stream. Only work when `sync_op` is true.

对应中文文档，https://github.com/PaddlePaddle/docs/pull/5225

### CODE REVIEW

File: paddle/fluid/distributed/collective/ProcessGroup.h [hunk 1]
```diff
@@ -90,6 +90,15 @@ class ProcessGroup {
         "ProcessGroup%s does not support allreduce", GetBackendName()));
   }
 
+  virtual void AllReduceOnCalcStream(
```
Comment (gongweibao) [inline paddle/fluid/distributed/collective/ProcessGroup.h]: Do we need a `StreamProcessGroup` base class?
`ProcessGroup` doesn't need to export a `stream` concept.
And ProcessGroupNCCL can inherit from `StreamProcessGroup`.

File: python/paddle/distributed/communication/stream/all_reduce.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            if local_rank == 0:
+                data = paddle.to_tensor([[4, 5, 6], [4, 5, 6]])
+            else:
+                data = paddle.to_tensor([[1, 2, 3], [1, 2, 3]])
+            task = dist.stream.all_reduce(data, sync_op=False)
+            task.wait()
+            out = data.numpy()
+            # [[5, 7, 9], [5, 7, 9]]
+    """
+    if group is not None and not group.is_member():
```
Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py:75]: Assert or raise an exception?

File: python/paddle/distributed/communication/stream/all_reduce.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import paddle.fluid.framework as framework
+from ...collective import _get_default_group, _get_reduce_op, ReduceOp
+
+
+def _all_reduce_in_dygraph(tensor, op, group, sync_op, use_calc_stream):
+    op_type = _get_reduce_op(op, "all_reduce")
+    group = _get_default_group() if group is None else group
+    if use_calc_stream:
+        group.process_group.allreduce_on_calc_stream(tensor, op_type)
+        return
```
Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py]: return a fake task?

File: python/paddle/distributed/communication/stream/all_reduce.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    task = group.process_group.allreduce(tensor, op_type)
+    if sync_op:
+        task.wait()
+
+    return None if sync_op else task
```
Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py]: The return value should be thought about again.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        finally:
+            if self._save_log_dir:
+                temp_log_dir_name = os.path.basename(self._log_dir.name)
+                dir_name = os.path.join(self._save_log_dir, temp_log_dir_name)
+                if not os.path.isdir(dir_name):
+                    print("The running logs will copy to {}".format(dir_name))
+                    shutil.copytree(self._log_dir.name, dir_name)
+
+
+def gen_envs_list(default_envs, changeable_envs):
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py]: `gen_product_envs_list`

File: python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                .format(err.cmd, err.timeout))
+        except subprocess.CalledProcessError as err:
+            raise RuntimeError(
+                "Error occurs when running this test case. The return code of command {} is {}"
+                .format(err.cmd, err.returncode))
+        finally:
+            if self._save_log_dir:
+                temp_log_dir_name = os.path.basename(self._log_dir.name)
+                dir_name = os.path.join(self._save_log_dir, temp_log_dir_name)
+                if not os.path.isdir(dir_name):
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py]: Assert that this log does not exists? And what we should do when it exists?

File: python/paddle/distributed/communication/stream/all_reduce.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def all_reduce(tensor,
+               op=ReduceOp.SUM,
+               group=None,
+               sync_op=True,
+               use_calc_stream=False):
+    """
+
+    After gather the input from all ranks in the group, do specific reduce operation on the gather result.
```
Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py]: An example：
```
The AllReduce operation is performing reductions on data (for example, sum, max) across devices and writing the result in the receive buffers of every rank.
```
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/operations.html

File: python/paddle/distributed/communication/stream/all_reduce.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    Args:
+        tensor (Tensor): The input tensor on each rank. The result will overwrite this tenor after communication. Support
+            float16, float32, float64, int32 or int64 as the input data type.
+        op (ReduceOp.SUM|ReduceOp.MAX|ReduceOp.Min|ReduceOp.PROD, optional): The reduce operation used. If none is given, use ReduceOp.SUM as default.
+        group (Group, optional): Communicate in which group. If none is given, use the global group as default.
+        sync_op (bool, optional): Indicate whether the communication is sync or not. If none is given, use true as default.
+        use_calc_stream (bool, optional): Indicate whether the communication is done on calculation stream. If none is given, use false as default. This
+            option is design for high performance demand, be careful to turn it on except you are clearly know its meaning.
```
Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py]: Grammarly can help you check grammar of them.

Comment (gongweibao) [inline python/paddle/distributed/communication/stream/all_reduce.py]: designed

**Review by gongweibao — CHANGES_REQUESTED** [review]
Some comments.

File: python/paddle/distributed/communication/stream/all_reduce.py (hunk 8 above)
Comment (gongweibao) [reply python/paddle/distributed/communication/stream/all_reduce.py]: https://www.grammarly.com/

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.h [hunk 9]
```diff
@@ -89,6 +95,17 @@ class ProcessGroupNCCL : public ProcessGroup {
     return std::string(NCCL_BACKEND_NAME);
   }
 
+  std::shared_ptr<ProcessGroup::Task> AllReduce(
+      std::vector<phi::DenseTensor>& in_tensors,   // NOLINT
```
Comment (gongweibao) [inline paddle/fluid/distributed/collective/ProcessGroupNCCL.h:100]: in_tensors will be modified?

File: paddle/fluid/distributed/collective/ProcessGroup.h (hunk 1 above)
Comment (LiYuRio) [reply paddle/fluid/distributed/collective/ProcessGroup.h]: Add a class named ProcessGroupStream

File: python/paddle/distributed/communication/stream/all_reduce.py (hunk 2 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/all_reduce.py:75]: done

File: python/paddle/distributed/communication/stream/all_reduce.py (hunk 3 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/all_reduce.py]: done

File: python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py (hunk 5 above)
Comment (LiYuRio) [reply python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py]: done

File: python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py (hunk 6 above)
Comment (LiYuRio) [reply python/paddle/fluid/tests/unittests/collective/test_communication_api_base.py]: move this to tearDown will solve this problem

File: python/paddle/distributed/communication/stream/all_reduce.py (hunk 8 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/all_reduce.py]: done

File: python/paddle/distributed/communication/stream/all_reduce.py (hunk 7 above)
Comment (LiYuRio) [reply python/paddle/distributed/communication/stream/all_reduce.py]: done

**Review by gongweibao — APPROVED** [review]
LGTM

File: paddle/fluid/distributed/collective/ProcessGroupNCCL.h (hunk 9 above)
Comment (LiYuRio) [reply paddle/fluid/distributed/collective/ProcessGroupNCCL.h:100]: this will be modified in the following pr

**Review by liuTINA0907 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
LGTM 单测时间设置

**Pull request merged**